Documentation/documentation.org

```diff
@@ -0,0 +1,259 @@
+
+# # ~:chart-type~ argument allows to specify a desired chart. Function
+# # accepts other keyword arguments which depend on a chart type. Range
+# # for X and Y axis in this example are inferred automatically but in
+# # principle each serie may have its own scale, scale and step (for
+# # functions).
+
+# # Another typical use of the library is to plot a function. We need to
+# # specify the X range and step.
+
+
+
+# * Embedding in a CLIM application
+# ** As a frame
+# ** As a pane
+# ** As an output record
+   
+# * Extending Polyclot
+
+* Reference manual
+** Data Frames
+*** Types
+**** <data-frame>                                                     :class:
+A protocol class. All class implementing this protocol must have it as
+its superclass.
+**** <raw-data-frame>                                                 :class:
+Data frame created by make-data-frame.
+**** %col-index%                                                       :type:
+Either an integer from 0 to n-1 or a string. If there is no such
+column an error <invalid-index> is signalled.
+**** %row-index%                                                       :type:
+Either an Integer from 0 to n-1 or a sequence.
+**** %col-slice%                                                       :type:
+
+- (eql t) :: select all rows/cols
+- (cons %col-index% %col-index%) :: select elements between indexes
+- (cons (eql t) %col-index%) :: select all elements up to the index
+- (cons %col-index% (eql t)) :: select all elements starting from the index
+- (vector %col-index%) :: select elements with specified indexes
+
+**** %row-slice%                                                       :type:
+
+- (eql t) :: select all rows/cols
+- (cons %row-index% %row-index%) :: select elements between indexes
+- (cons (eql t) %row-index%) :: select all elements up to the index
+- (cons %row-index% (eql t)) :: select all elements starting from the index
+- (vector %row-index%) :: select elements with specified indexes
+**** %row-spec%                                                        :type:
+- (or %row-index% row) :: used in map-data-frame-cols
+**** <invalid-slice> (error)                                      :condition:
+**** <invalid-index> (error)                                      :condition:
+**** <row-invalid-index> (<invalid-index>)                        :condition:
+**** <col-invalid-index> (<invalid-index>)                        :condition:
+**** <row-does-not-exist> (<row-invalid-index>)                   :condition:
+**** <col-does-not-exist> (<col-invalid-index>)                   :condition:
+**** <insert-error> (error)                                       :condition:
+**** <col-name-not-unique> (<insert-error>)                       :condition:
+**** <row-length-mismatch> (<insert-error>)                       :condition:
+*** Accessors                                                            :gf:
+**** dims <data-frame>
```

Comment:
Why two values and not a cons or something like it, more similar to array-dimensions?

Reply:
I didn't think about it really. It avoids consing I suppose. Should I change it?

Reply:
That would be my suggestion. Unless I am thinking about it incorrectly, with the second dimension being the second value, you would always need to do a multiple-value-bind. As a cons, it is more like the `#'array-dimensions`.

Reply:
getting list is simple: (multiple-value-list (dims df)), if you want to access a specific value, you may call (nth-value 1 (dims df))

Reply:
Quite right. Using multiple values always requires a wrapper around the function that understands multiple values, which is why I generally avoid it, but this is perhaps a wrongly based bias. Some (pjb) have suggested that "always use multiple values!". So it is!